dvc/dependency/__init__.py

```diff
@@ -87,26 +89,31 @@ def loads_from(stage, s_list, erepo=None):
     return ret
 
 
-def _parse_params(path_params):
-    path, _, params_str = path_params.rpartition(":")
-    params = params_str.split(",")
-    return path, params
+def _merge_params(s_list):
+    d = defaultdict(list)
+    for key in s_list:
+        if isinstance(key, str):
+            path = ParamsDependency.DEFAULT_PARAMS_FILE
+            params = [key]
+        else:
+            if not isinstance(key, dict):
+                msg = "Only list of str/dict is supported. Got: "
+                msg += f"'{type(key).__name__}'."
+                raise ValueError(msg)
+            path = first(key)
+            if not path:
+                continue
+            params = key[path]
+            if not isinstance(params, list):
+                msg = "Expected list of params for custom params file "
+                msg += f"'{path}', got '{type(params).__name__}'."
+                raise ValueError(msg)
```

Comment:
Didn't want to add assertions, so, I have used ValueError. They are purely internal, as error are likely to be raised by schema validators before even reaching here.